In [9]:
import pandas as pd

df = pd.read_csv("../data/medical-appointments-no-show-en.csv")
df.head(10)


,specialty,appointment_time,gender,appointment_date,no_show,no_show_reason,disability,date_of_birth,entry_service_date,city,...,over_60_years_old,patient_needs_companion,average_temp_day,average_rain_day,max_temp_day,max_rain_day,rainy_day_before,storm_day_before,rain_intensity,heat_intensity
0,physiotherapy,13:20,M,09/09/2021,yes,surto,NaN,NaN,NaN,NaN,...,0,0,20.75,0.01,23.7,0.2,1,1,no_rain,mild
1,psychotherapy,13:20,M,09/09/2021,no,NaN,NaN,NaN,NaN,NaN,...,0,0,20.75,0.01,23.7,0.2,1,1,no_rain,mild
2,speech therapy,13:20,F,09/09/2021,no,NaN,NaN,NaN,NaN,NaN,...,0,0,20.75,0.01,23.7,0.2,1,1,no_rain,mild
3,physiotherapy,13:20,F,09/09/2021,no,NaN,NaN,NaN,NaN,NaN,...,0,0,20.75,0.01,23.7,0.2,1,1,no_rain,mild
4,physiotherapy,14:00,M,09/09/2021,no,NaN,motor,10/10/1954,5/2/2020,B. CAMBORIU,...,1,1,20.75,0.01,23.7,0.2,1,1,no_rain,mild
5,occupational therapy,14:00,M,09/09/2021,no,NaN,intellectual,21/3/2017,26/11/2019,ITAJAÍ,...,0,1,20.75,0.01,23.7,0.2,1,1,no_rain,mild
6,psychotherapy,14:00,M,09/09/2021,no,NaN,intellectual,10/1/2011,1/10/2019,CAMBORIU,...,0,1,20.75,0.01,23.7,0.2,1,1,no_rain,mild
7,psychotherapy,14:00,F,09/09/2021,no,NaN,NaN,NaN,NaN,NaN,...,0,0,20.75,0.01,23.7,0.2,1,1,no_rain,mild
8,occupational therapy,14:00,M,09/09/2021,no,NaN,intellectual,21/3/2017,20/11/2019,ITAJAÍ,...,0,1,20.75,0.01,23.7,0.2,1,1,no_rain,mild
9,speech therapy,14:00,F,09/09/2021,no,NaN,NaN,NaN,NaN,NaN,...,0,0,20.75,0.01,23.7,0.2,1,1,no_rain,mild


In [10]:
df.shape

(49593, 26)

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 49593 entries, 0 to 49592
Data columns (total 26 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   specialty                42139 non-null  str    
 1   appointment_time         49593 non-null  str    
 2   gender                   49593 non-null  str    
 3   appointment_date         49593 non-null  str    
 4   no_show                  49593 non-null  str    
 5   no_show_reason           1737 non-null   str    
 6   disability               44456 non-null  str    
 7   date_of_birth            39272 non-null  str    
 8   entry_service_date       44438 non-null  str    
 9   city                     44412 non-null  str    
 10  icd                      10717 non-null  str    
 11  appointment_month        49593 non-null  str    
 12  appointment_year         49593 non-null  int64  
 13  appointment_shift        49593 non-null  str    
 14  age                      39243 no

In [12]:
df["no_show"].value_counts()

no_show
no     44761
yes     4832
Name: count, dtype: int64

In [13]:
df.isna().sum().sort_values(ascending=False)

no_show_reason             47856
icd                        38876
age                        10350
date_of_birth              10321
specialty                   7454
city                        5181
entry_service_date          5155
disability                  5137
max_rain_day                1016
max_temp_day                1016
average_rain_day            1016
average_temp_day            1016
patient_needs_companion        0
rain_intensity                 0
storm_day_before               0
rainy_day_before               0
appointment_shift              0
over_60_years_old              0
under_12_years_old             0
appointment_time               0
appointment_year               0
appointment_month              0
no_show                        0
appointment_date               0
gender                         0
heat_intensity                 0
dtype: int64

In [14]:
df["no_show"].value_counts()

no_show
no     44761
yes     4832
Name: count, dtype: int64

In [15]:
df["no_show"].value_counts(normalize=True) * 100

no_show
no     90.256689
yes     9.743311
Name: proportion, dtype: float64

In [16]:
df.isna().sum().sort_values(ascending=False)

no_show_reason             47856
icd                        38876
age                        10350
date_of_birth              10321
specialty                   7454
city                        5181
entry_service_date          5155
disability                  5137
max_rain_day                1016
max_temp_day                1016
average_rain_day            1016
average_temp_day            1016
patient_needs_companion        0
rain_intensity                 0
storm_day_before               0
rainy_day_before               0
appointment_shift              0
over_60_years_old              0
under_12_years_old             0
appointment_time               0
appointment_year               0
appointment_month              0
no_show                        0
appointment_date               0
gender                         0
heat_intensity                 0
dtype: int64

In [17]:
df.nunique().sort_values()

storm_day_before              2
rainy_day_before              2
no_show                       2
patient_needs_companion       2
over_60_years_old             2
under_12_years_old            2
appointment_shift             2
gender                        3
disability                    3
rain_intensity                4
heat_intensity                5
appointment_year              7
specialty                     8
appointment_month            12
city                         13
icd                          63
appointment_time             74
max_rain_day                 84
age                          87
average_rain_day            130
max_temp_day                185
entry_service_date          333
no_show_reason              341
average_temp_day            677
date_of_birth               954
appointment_date           1001
dtype: int64

In [18]:
for col in [
    "gender",
    "appointment_shift",
    "disability",
    "rain_intensity",
    "heat_intensity",
    "specialty",
    "city"
]:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False))


--- gender ---
gender
M    37583
F    12003
I        7
Name: count, dtype: int64

--- appointment_shift ---
appointment_shift
afternoon    27865
morning      21728
Name: count, dtype: int64

--- disability ---
disability
intellectual    30882
motor           13565
NaN              5137
                    9
Name: count, dtype: int64

--- rain_intensity ---
rain_intensity
no_rain     35362
moderate     6985
weak         4802
heavy        2444
Name: count, dtype: int64

--- heat_intensity ---
heat_intensity
mild          23413
cold          14293
warm           9890
heavy_cold     1486
heavy_warm      511
Name: count, dtype: int64

--- specialty ---
specialty
speech therapy          11877
psychotherapy           11675
physiotherapy            9074
NaN                      7454
occupational therapy     6486
pedagogo                 2064
enf                       712
assist                    194
sem especialidade          57
Name: count, dtype: int64

--- city ---
city
ITAJAÍ            

In [20]:
def no_show_rate_by(column):
    result = (
        df.groupby(column, dropna=False)["no_show"]
        .apply(lambda x: (x == "yes").mean() * 100)
        .sort_values(ascending=False)
        .round(2)
    )
    return result

In [21]:
no_show_rate_by("appointment_shift")

appointment_shift
afternoon    9.91
morning      9.53
Name: no_show, dtype: float64

In [22]:
no_show_rate_by("rain_intensity")

rain_intensity
heavy       10.47
moderate    10.35
weak         9.85
no_rain      9.56
Name: no_show, dtype: float64

In [23]:
no_show_rate_by("heat_intensity")

heat_intensity
heavy_cold    14.33
heavy_warm    11.35
warm          11.03
cold          10.44
mild           8.45
Name: no_show, dtype: float64

In [24]:
no_show_rate_by("specialty")

specialty
sem especialidade       22.81
NaN                     12.09
physiotherapy           10.37
speech therapy           9.44
psychotherapy            9.34
occupational therapy     8.56
pedagogo                 7.46
assist                   6.70
enf                      6.04
Name: no_show, dtype: float64

In [25]:
def no_show_summary_by(column):
    summary = (
        df.groupby(column, dropna=False)["no_show"]
        .agg(
            total_appointments="count",
            no_show_count=lambda x: (x == "yes").sum(),
            no_show_rate=lambda x: (x == "yes").mean() * 100
        )
        .sort_values("no_show_rate", ascending=False)
    )
    
    summary["no_show_rate"] = summary["no_show_rate"].round(2)
    return summary

In [26]:
no_show_summary_by("specialty")

,total_appointments,no_show_count,no_show_rate
specialty,,,
sem especialidade,57,13,22.81
NaN,7454,901,12.09
physiotherapy,9074,941,10.37
speech therapy,11877,1121,9.44
psychotherapy,11675,1091,9.34
occupational therapy,6486,555,8.56
pedagogo,2064,154,7.46
assist,194,13,6.70
enf,712,43,6.04


In [27]:
no_show_summary_by("heat_intensity")

,total_appointments,no_show_count,no_show_rate
heat_intensity,,,
heavy_cold,1486,213,14.33
heavy_warm,511,58,11.35
warm,9890,1091,11.03
cold,14293,1492,10.44
mild,23413,1978,8.45


In [28]:
no_show_summary_by("rain_intensity")

,total_appointments,no_show_count,no_show_rate
rain_intensity,,,
heavy,2444,256,10.47
moderate,6985,723,10.35
weak,4802,473,9.85
no_rain,35362,3380,9.56


### Initial Observations

- The target variable `no_show` is imbalanced: approximately 9.7% of appointments are no-shows.
- Appointments on heavy cold days show a noticeably higher no-show rate (14.33%) compared to mild days (8.45%).
- Rain intensity appears to be associated with a slight increase in no-show rates, from 9.56% on no-rain days to 10.47% on heavy-rain days.
- Missing `specialty` information is associated with a relatively high no-show rate (12.09%), which may indicate a data quality or operational pattern worth investigating.

In [29]:
no_show_summary_by("disability")

,total_appointments,no_show_count,no_show_rate
disability,,,
,9,4,44.44
NaN,5137,778,15.15
motor,13565,1308,9.64
intellectual,30882,2742,8.88


In [30]:
no_show_summary_by("patient_needs_companion")

,total_appointments,no_show_count,no_show_rate
patient_needs_companion,,,
0,22603,2416,10.69
1,26990,2416,8.95


In [31]:
no_show_summary_by("under_12_years_old")

,total_appointments,no_show_count,no_show_rate
under_12_years_old,,,
0,26163,2772,10.60
1,23430,2060,8.79


In [32]:
no_show_summary_by("over_60_years_old")

,total_appointments,no_show_count,no_show_rate
over_60_years_old,,,
1,3560,356,10.00
0,46033,4476,9.72


In [33]:
no_show_summary_by("gender")

,total_appointments,no_show_count,no_show_rate
gender,,,
F,12003,1240,10.33
M,37583,3592,9.56
I,7,0,0.00


In [34]:
no_show_summary_by("city")

,total_appointments,no_show_count,no_show_rate
city,,,
MONTENEGRO,27,6,22.22
BLUMENAU,16,3,18.75
NaN,5181,784,15.13
LUIZ ALVES,523,71,13.58
ILHOTA,583,77,13.21
BOMBINHAS,1707,200,11.72
PORTO BELO,950,111,11.68
NAVEGANTES,3901,410,10.51
ITAPEMA,2665,258,9.68


## Initial Data Understanding - Key Takeaways

- The dataset contains **49,593 medical appointments** and **26 variables**.
- The target variable is `no_show`, where approximately **9.74%** of appointments are no-shows.
- This is an **imbalanced classification problem**, so accuracy alone will not be sufficient during model evaluation.

### Early patterns observed
- No-show rate is higher on **heavy cold days** (**14.33%**) compared to **mild days** (**8.45%**).
- Rain intensity shows a smaller but gradual increase in no-show rates, from **9.56%** on no-rain days to **10.47%** on heavy-rain days.
- Appointments with missing `specialty`, `disability`, or `city` information show noticeably higher no-show rates.
- Patients under 12 years old have a lower no-show rate (**8.79%**) compared to other patients (**10.60%**).
- Patients who do not need a companion show a somewhat higher no-show rate (**10.69%**) than those who do (**8.95%**).
- Female patients show a slightly higher no-show rate (**10.33%**) than male patients (**9.56%**).
- City-level no-show rates vary, but small sample sizes should be interpreted carefully.